### PR TITLE
 Use region values in SplitMulti.

### PR DESCRIPTION
**Pull request by cseed**
Generic: computation of split values is done in expression language.
Support HTS-like split of generic schema.
Removed fakeRef, which I don't think anyone ever used.  If anyone wants it, can be added through expression language once that is exposed.

@tpoterba FYI I removed fakeRef.

### CODE REVIEW

**Comment by cseed** [comment]
@tpoterba I also got rid of max_shift and (hopefully) made things more robust.  Can you take a quick look to see if you're happy with what I did?

**Comment by tpoterba** [comment]
I'm glad fakeref is gone, that was some BS.

File: src/main/scala/is/hail/expr/FunctionRegistry.scala [hunk 1]
```diff
@@ -858,6 +853,26 @@ object FunctionRegistry {
...
+  register("downcode", { (c: Call, i: Int) =>
+    val p = Genotype.gtPair(c)
+    Call((if (p.j == i) 1 else 0) +
+      (if (p.k == i) 1 else 0))
+  },
+    """
+    Downcode call by sending all alleles except i to the reference.
+    """,
+    "gt" -> "genotype call to be downcoded",
+    "i" -> "allele to become the referene allele")(callHr, int32Hr, callHr)
```
Comment (catoverdrive) [inline src/main/scala/is/hail/expr/FunctionRegistry.scala]: *reference

File: src/main/scala/is/hail/methods/SplitMulti.scala [hunk 2]
```diff
@@ -1,158 +1,213 @@
...
-        (v, (insertSplitAnnots(va, ind, true), gsb.result()))
-      }
-  }
+    val va = ur.get(2)
 
-  def splitNumber(str: String): String =
-    if (str == "A" || str == "R" || str == "G")
-      "."
-    else str
+    if (v.isBiallelic) {
```
Comment (catoverdrive) [inline src/main/scala/is/hail/methods/SplitMulti.scala]: you mentioned on Thursday that the fact that gAnnotator is ignored in the biallelic might pose a problem if people were to write their own expressions for the row-splitting. Does that make sense to change though if the point of the expressions is to provide a way to split the rows and biallelic rows don't need to be split?

File: src/main/scala/is/hail/methods/SplitMulti.scala [hunk 3]
```diff
@@ -1,158 +1,213 @@
...
+          newMatrixType.orderedRDD2Type,
+          vsm.rdd2.orderedPartitioner,
+          split(sortAlleles = true, removeLeftAligned = false, removeMoving = true, verifyLeftAligned = false))
+
+        val movedVariants = OrderedRDD2.shuffle(
+          newMatrixType.orderedRDD2Type,
+          vsm.rdd2.orderedPartitioner,
+          split(sortAlleles = false, removeLeftAligned = true, removeMoving = false, verifyLeftAligned = false))
+
+        leftAlignedVariants.partitionSortedUnion(movedVariants)
```
Comment (catoverdrive) [inline src/main/scala/is/hail/methods/SplitMulti.scala]: so this creates 2 separate rdds of sorted & unsorted elements, orders the second one, and then creates a sorted union?

Comment (cseed) [reply src/main/scala/is/hail/methods/SplitMulti.scala]: Correct!  The goal is to minimize the size of the shuffled data by only shuffling the things that moved and are therefore (potentially) no longer ordered.  This is in my mind a major improvement over the previous code that tried to re-sort partitions and shuffle only things that moved between partitions.  Re-sorting partitions is really a no-go unless you support off-disk shuffling (as shuffle does) and was a source of memory/failure problems previously that I think this should address.

File: src/main/scala/is/hail/methods/SplitMulti.scala (hunk 2 above)
Comment (cseed) [reply src/main/scala/is/hail/methods/SplitMulti.scala]: That's correct.  I removed the special case handling of biallelics.  I think we should look at the performance once all the region value stuff is working nicely and then decide if it is necessary to special case biallelics again.

**Comment by cseed** [comment]
Yes, there was a conflict.  I've resolved it and tests are running now.

**Pull request merged**